test/src/main/java/org/springframework/security/test/context/support/WithMockOidcUser.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2002-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.test.context.support;
+
+import org.springframework.core.annotation.AliasFor;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.context.SecurityContext;
+import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken;
+import org.springframework.test.context.TestContext;
+import org.springframework.test.web.servlet.MockMvc;
+
+import java.lang.annotation.*;
+
+/**
+ * When used with {@link WithSecurityContextTestExecutionListener} this annotation can be
+ * added to a test method to emulate running with a mocked user. In order to work with
+ * {@link MockMvc} The {@link SecurityContext} that is used will have the following
+ * properties:
+ *
+ * <ul>
+ * <li>The Authentication that is populated in the {@link SecurityContext} is of type {@link OAuth2AuthenticationToken}.</li>
+ * <li>The principal on the Authentication is Spring Security’s User object of type {@code OidcUser}.</li>
+ * <li>The default User has the username of "user", {@link #value()} or {@link #name()} and does not have to exist.</li>
+ * <li>The default User has and a single GrantedAuthority {@link GrantedAuthority} or those that are specified by {@link #authorities()}.</li>
+ * </ul>
+ *
+ * @author Nena Raab
+ * @see WithUserDetails
+ * @since 5.3
+ */
+@Target({ ElementType.METHOD, ElementType.TYPE })
+@Retention(RetentionPolicy.RUNTIME)
+@Inherited
+@Documented
+@WithSecurityContext(factory = WithMockOidcUserSecurityContextFactory.class)
+public @interface WithMockOidcUser {
+	String DEFAULT_SCOPE = "SCOPE_openid";
+
+	/**
+	 * Convenience mechanism for specifying the username. The default is "user". If
+	 * {@link #name()} is specified it will be used instead of {@link #value()}
+	 * @return
+	 */
+	String value() default "user";
+
+	/**
+	 * The user name oder user id (subject) to be used. Note that {@link #value()} is a synonym for
+	 * {@link #name()}, but if {@link #name()} is specified it will take
+	 * precedence.
+	 * @return
+	 */
+	String name() default "";
+
+	/**
+	 * <p>
+	 * The authorities, that should be mapped to {@code GrantedAuthority}.
+	 * The default is "SCOPE_openid". A {@link GrantedAuthority} will be created for each value.
+	 * </p>
+	 **
+	 * @return
+	 */
+	String[] authorities() default { "SCOPE_openid" };
```

Comment:
My suggestion would be to use scopes instead of authorities, as authorities has Spring Security specific means (gets prefixed with "SCOPE_") and i can not see an advantage of having both, beside of the fact when someone has changed the prefix.

Reply:
No problem, I don't have strong feelings about it, and we can always wait for a community member to request it.

The rest is just for future reference:

> mutual exclusivity

Btw, while I see your point about mutual exclusivity, I don't see it as a problem as you do. The DSL allows `oauth2ResourceServer().jwt()` OR `oauth2ResourceServer().opaqueToken()`, for example.

> scopes only

agreed

> authorities only ... add "SCOPE_" prefix

The application only has to do this if it's using the "SCOPE_" prefix in its authorization rules, in which case the tester already knows about it. I don't see this as an issue.

Reply:
then - lets provide both :-)